sdk/core/azure-core/test/ut/retry_policy.cpp

```diff
@@ -0,0 +1,324 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "http/retry_policy_private.hpp"
+
+#include <gtest/gtest.h>
+
+using namespace Azure::Core::Http;
+using namespace Azure::Core::Http::Policies;
+using namespace Azure::Core::Http::Policies::_detail;
+
+TEST(RetryPolicy, Exponential)
+{
+  using namespace std::chrono_literals;
+
+  RetryOptions const options{3, 1s, 2min, {}};
+
+  {
+    std::chrono::milliseconds retryAfter{};
+    bool const shouldRetry = ShouldRetryOnTransportFailure(options, 1, retryAfter, 1.0);
+    EXPECT_EQ(shouldRetry, true);
+    EXPECT_EQ(retryAfter, 1s);
+  }
+
+  {
+    std::chrono::milliseconds retryAfter{};
+    bool const shouldRetry = ShouldRetryOnTransportFailure(options, 2, retryAfter, 1.0);
+    EXPECT_EQ(shouldRetry, true);
+    EXPECT_EQ(retryAfter, 2s);
+  }
+
+  {
+    std::chrono::milliseconds retryAfter{};
+    bool const shouldRetry = ShouldRetryOnTransportFailure(options, 3, retryAfter, 1.0);
+    EXPECT_EQ(shouldRetry, true);
+    EXPECT_EQ(retryAfter, 4s);
+  }
+
+  {
+    std::chrono::milliseconds retryAfter{};
+    bool const shouldRetry = ShouldRetryOnTransportFailure(options, 4, retryAfter, 1.0);
+    EXPECT_EQ(shouldRetry, false);
```

Comment:
should we test that `retryAfter` is not updated when `false` is returned?

Reply:
I don't think so. The value should remain untouched, but in general, this is basically a private function, and I have an expectation that noone is attempting to use `retryAfter` if the function returned `false`. This, theoretically, can change in the future, but right now, noone should have any expectations. Today, I would not write any code that spends any cycles taking care of any default value for the out parameter when `false` is returned, or, the opposite, that makes sure that the out value does not get touched if we're about to return false. This is a basically a private method with its own limited set of expectations and preconditions.